Justin/favoriting

thisjustin123 opened the pull request:
Implemented favoriting and a rudimentary Proto Datastore implementation for storing favorite data (only favoriting, for now).

Changes:
- Star icon on EateryCard now allows users to favorite eateries.
- Star button on EateryDetailScreen now allows users to favorite eateries.
- Favorited eateries will appear under the "Favorite Eateries" section in Home.
- When eateries are favorited in Home, the "Favorite Eateries" tab will automatically add or remove newly favorited/unfavorited eateries.
- Added ability to observe any eateries' favorite statuses live with MutableStates.
- Favorited eateries will locally save + load.

Most of the above functionality was put in a new utilities file, FavoritingUtils.kt

![Screenshot_1646288845](https://user-images.githubusercontent.com/47724806/156509116-b64cb6f9-5a5f-4892-8fc6-55d32ee1640b.png)

kevinsun-dev commented on app/build.gradle:
```diff
@@ -69,6 +70,10 @@ dependencies {
+    implementation 'com.google.code.gson:gson:2.8.7'
+    implementation  "androidx.datastore:datastore-core:1.0.0"
+    implementation  "com.google.protobuf:protobuf-javalite:3.18.0"
+    implementation("androidx.datastore:datastore-preferences:1.0.0")
```
I love how there's 3 different coding conventions across all 4 lines. 😆 
Line 73: Single Quotes
Line 74-75: Double Quotes, extra space before
Line 76: Parenthesis, double quotes, no space.

Can we change these to:
```
implementation "com.google.code.gson:gson:2.8.7"
implementation "androidx.datastore:datastore-core:1.0.0"
implementation "com.google.protobuf:protobuf-javalite:3.18.0"
implementation "androidx.datastore:datastore-preferences:1.0.0"
```

kevinsun-dev commented on app/src/main/java/com/appdev/eateryblueandroid/util/FavoritingUtils.kt:
```diff
@@ -0,0 +1,106 @@
+    if (mPreferencesEditor != null) return mPreferencesEditor.commit()
+    return false
+}
+*/
```
If it's no longer necessary, you can remove the commented code. Same goes for ln 18 & ln 20 of this file.

kevinsun-dev reviewed (approved):
Very nice first commit, well documented too. I see very little that needs changing aside from a few syntactic details.

thisjustin123 commented:
Found a bug that causes a disconnect between MutableStates and global map--will push out a commit to fix soon.

aasthashah999 commented on app/build.gradle:
```diff
@@ -77,4 +82,23 @@ dependencies {
+        artifact = "com.google.protobuf:protoc:3.14.0"
+    }
+
+    // Generates the java Protobuf-lite code for the Protobufs in this project. See
```
We can remove these comments

aasthashah999 reviewed (approved):
Overall looks great! Great job :) Just remove some comments and fix styling in the dependencies

thisjustin123 commented:
Okay, that commit I just pushed fixed the issue. For some reason some of my previous changes to HomeViewModel.kt did not save, so I re-added them.

connorreinhold commented on app/src/main/java/com/appdev/eateryblueandroid/util/FavoritingUtils.kt:
```diff
@@ -0,0 +1,106 @@
+    return state
+}
+
+fun numFavorites(): Int {
```
I think should be in separate file (think it is better to separate files by what they are doing, rather than the feature they represent).

honestly don't think this function will be used too much so we can just put it in the file that used it.

connorreinhold reviewed (commented):
Looks dope! Great that u got it working so fast. Just have a few comments on the architecture of how we represent "truth" for whether an eatery is a favorited eatery or not

thisjustin123 commented on app/src/main/java/com/appdev/eateryblueandroid/util/FavoritingUtils.kt:
```diff
@@ -0,0 +1,106 @@
+private const val DATA_STORE_FILE_NAME = "user_prefs.pb"
+//private const val SORT_ORDER_KEY = "sort_order"
+
+private var myFavorites: Favorites = Favorites(mutableMapOf())
```
I can definitely rewrite the code to get rid of the Favorites data class. It was an artifact left over from when I implemented save/load from SharedPreferences, so it is unnecessary.

thisjustin123 commented on app/src/main/java/com/appdev/eateryblueandroid/util/FavoritingUtils.kt:
```diff
@@ -0,0 +1,106 @@
+fun toggleFavorite(eatery: Eatery) {
+    eatery.isFavorite = !eatery.isFavorite
+    favoriteStates[eatery.id]!!.value = !favoriteStates[eatery.id]!!.value
+    myFavorites.favoritesMap[eatery.id] = eatery.isFavorite
```
Good point--I will rewrite to include only favoriteStates. favoritesMap is not really needed, for everything it does can also be done with favoriteStates, and more.

thisjustin123 commented on app/src/main/java/com/appdev/eateryblueandroid/util/FavoritingUtils.kt:
```diff
@@ -0,0 +1,106 @@
+
+private var myFavorites: Favorites = Favorites(mutableMapOf())
+var appContext: Context? = null
+private var favoriteStates : HashMap<Int, MutableState<Boolean>> = hashMapOf()
```
I attempted to structure favoriting like this in the beginning, but ran into problems when attempting to use eatery.isFavorite to update displays on EateryCard. I had to switch to a MutableState structure to fix this. eatery.isFavorite should still work in any context other than live data observation due to the lines I wrote in HomeViewModel.kt

thisjustin123 commented on app/src/main/java/com/appdev/eateryblueandroid/util/FavoritingUtils.kt:
```diff
@@ -0,0 +1,106 @@
+            }
+
+        favoritesFlow.collect { map -> myFavorites.favoritesMap = map.toMutableMap() }
+    }
```
No, I could not find a way to "pass" the list of eateries being used by the app to FavoritingUtils.kt so that they could be updated here. My lines in HomeViewModel.kt initialize eatery.isFavorite, though.

thisjustin123 commented on app/src/main/java/com/appdev/eateryblueandroid/util/FavoritingUtils.kt:
```diff
@@ -0,0 +1,106 @@
+    return num
+}
+
+fun isFavorite(eatery: Eatery): Boolean {
```
eatery.isFavorite should still work in any context other than live data observation--for that, use the MutableStates.

thisjustin123 commented:
Implemented @connorreinhold 's suggestions in the latest commit:
Regular favoriting functions (i.e. toggling favorite, getting isFavorite) have been moved to Eatery objects and removed from the utils file.

Examples of how setting/getting favorites has changed (Before -> After):
- isFavorite(eatery) -> eatery.isFavorite()
- getMutableStateOf(eatery).value -> eatery.isFavorite()
  (Above recomposes just fine whenever favorite status is changed!!)
- toggleFavorite(eatery) -> eatery.toggleFavorite()

These changes should make working with favoriting much more straightforward in other files.

connorreinhold commented on app/src/main/java/com/appdev/eateryblueandroid/models/Eatery.kt:
```diff
@@ -25,8 +27,32 @@ data class Eatery(
+) {
+    fun toggleFavorite() {
+        isFavorite.value = !(isFavorite.value)
+        com.appdev.eateryblueandroid.util.saveFavoriteMap()
```
can we import saveFavoriteMap?

connorreinhold commented on app/src/main/java/com/appdev/eateryblueandroid/util/FavoritingUtils.kt:
```diff
@@ -73,23 +51,27 @@ fun numFavorites(): Int {
     return num
```
can we remove numFavorites method and getMutableFavoriteStateOf method?

thisjustin123 replied:
I will remove getMutableFavoriteStateOf in my next changes, but I feel that keeping numFavorites is necessary since otherwise I would have to change favoriteMap to public, which runs the risk of breaking its invariant (all values in favoriteMap must be equal to eatery's .isFavorite() for all eateries) if future additions tamper with it unknowingly.

connorreinhold reviewed (changes requested):
Very close to what I was picturing. Sorry to be leaving so many comments but I think the following change will remove the problems that we are getting that made it necessary to add the `fetchedStates` and `hasSetFavorites` booleans. Is it necessary the pointers to these MutableState<Boolean> values in the FavoritingUtils class? Instead when we run initializeFavoritesMap, we can just update the mutableState(false) in the Eatery model to whatever the expected value is

Can we change `favoriteStates` to be a `HashMap<Int, Boolean>` and the `saveFavoriteMap()`method to `saveFavorite(eatery_id: Int, favorited: Boolean)`

thisjustin123 commented:
> Instead when we run initializeFavoritesMap, we can just update the mutableState(false) in the Eatery model to whatever the value we want is
> 
> Can we change `favoriteStates` to be a `HashMap<Int, Boolean>` and the `saveFavoriteMap()`method to `saveFavorite(eatery_id: Int, favorited: Boolean)`

Working on it, will push once this weird `HTTP 502 Bad Gateway` error goes away and lets me test it.

thisjustin123 commented:
Newest commit implements all requested changes.

connorreinhold reviewed (approved):
LGTM 🚀🚀
